ci: downgrade Chrome to fix flakes in aio e2e tests

gkalpak opened the pull request:
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] ~~Tests for the changes have been added (for bug fixes / features)~~
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~


## PR Type
```
[x] CI related changes
```

## What is the current behavior?
The angular.io e2e test were moved to an optional job, due to extreme flakiness.
Issue Number: #20159


## What is the new behavior?
ChromeDriver and Chrome are pinned to versions that do not cause flakes (2.32 and 59 respectively).
See #20159 and the commit messages for more info.


## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```


## Other information
Chrome(-ium) has been downgraded to v59 (which is compatible with ChromeDriver v2.32), so tests are not run against the latest version. I tested the current dev version of Chrome (v64) and there was still flakiness (which is expected if the issue is with ChromeDriver v2.33).

Fixes #20159.

##
**EDIT**:
This [Chromium bug][1] could be related. I have added info about our failures there.

[1]: https://bugs.chromium.org/p/chromedriver/issues/detail?id=2105

gkalpak committed:
test(aio): fix and clean up e2e tests

mary-poppins commented:
You can preview f050e38 at https://pr20403-f050e38.ngbuilds.io/.

mary-poppins commented:
You can preview 4dd870d at https://pr20403-4dd870d.ngbuilds.io/.

mary-poppins commented:
You can preview e8e2df0 at https://pr20403-e8e2df0.ngbuilds.io/.

mary-poppins commented:
You can preview 90b454e at https://pr20403-90b454e.ngbuilds.io/.

gkalpak committed:
ci(aio): move e2e tests to non-optional job

This essentially reverts #20178, since the flakes should be gone after
pinning ChromeDriver and Chrome versions to 2.32 and 59 respectively.

mary-poppins commented:
You can preview 712e2c1 at https://pr20403-712e2c1.ngbuilds.io/.

gkalpak commented:
The CircleCI linting issue is not related to this PR (comes from master).

petebacondarwin commented on aio/e2e/app.e2e-spec.ts:
```diff
@@ -42,24 +42,24 @@ describe('site App', function() {
 
-      page.navigateTo('api');
-      page.getScrollTop().then(scrollTop => expect(scrollTop).toBe(0));
+      page.getNavItem(/api/i).click();
```
Arguably this is a different test to before, no?
Previously, we are navigating by changing the URL but now by clicking on a link. The former is going to reload the page, right? While the latter will just update the doc being displayed.

gkalpak replied:
Yes, it is a different test. A correct test :smiley:
The previous test was broken, because it didn't actually test that we scroll to top when navigating to the `api` page. It did a full-page navigation (so of course `scrollTop` would be 0 :grin:)

gkalpak replied:
(FWIW, this had nothing to do with the flakes.)

petebacondarwin commented on aio/e2e/app.e2e-spec.ts:
```diff
@@ -42,24 +42,24 @@ describe('site App', function() {
 
-      page.navigateTo('guide/security');
-      page.getScrollTop().then(scrollTop => expect(scrollTop).toBe(0));
+      page.getNavItem(/security/i).click();
```
Same here.

gkalpak replied:
Same here :grin:

petebacondarwin commented on aio/e2e/app.po.ts:
```diff
@@ -27,11 +27,12 @@ export class SitePage {
       // See the GA snippet in index.html
-      .then(() => browser.executeScript('sessionStorage.setItem("__e2e__", true);'));
+      .then(() => browser.executeScript('sessionStorage.setItem("__e2e__", true);'))
+      .then(() => browser.waitForAngular());
```
Ooh, this is interesting. Is this required? Do we not wait for Angular by default?

gkalpak replied:
AFAICT, not all actions `waitForAngular`. ElementFinder actions do (which is what most people use n their tests), but `executeScript` doesn't. And we use quite a few `executeScript`s. Having the wait here is a convenience over having to do it in other places nd shouldn't incur any cost, because it will be called anyway eventually (directly or indirectly).

But having it here is not strictly required. Happy to remove it (and add it where it is necessary) if it feels too magic to you.

petebacondarwin replied:
No problem. I didn't realise that `executeScript` doesn't do that.

gkalpak replied:
Yeah, I didn't know either (until I looked through the code). I always assume that all protractor methods wait for Angular ¯\_(ツ)_/¯

IgorMinar reviewed (approved):
this looks good to me but now we have a new issue: how to get to the latest stable chrome? what needs to be done for that? is there an external issue that we should be tracking as a blocker for the upgrade?

petebacondarwin commented:
@gkalpak fancy some conflict resolution?

IgorMinar commented:
@gkalpak is this PR still relevant?

gkalpak commented:
Unfortunately yes. I've been looking through the issues on the ChromeDriver bug tracker and I think there are a couple that are related. I will finalize my investigation tomorrow and get this ready for merging.

IgorMinar commented:
ok. thanks

On Thu, Nov 30, 2017 at 3:43 PM George Kalpakas <notifications@github.com>
wrote:

> Unfortunately yes. I've been looking through the issues on the
> ChromeDriver bug tracker and I think there are a couple that are related. I
> will finalize my investigation tomorrow and get this ready for merging.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/20403#issuecomment-348357271>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANM6ObrbKOifYFnfdidpefZ21_4iTf1ks5s7z2ZgaJpZM4QcbaZ>
> .
>

IgorMinar commented:
@gkalpak  this still looks dead :-)

gkalpak commented:
Not sure what it means yet, but it seems that the breakage started somewehre between Chromium's revesion 494240 and 494248 (incl.).

gkalpak committed:
ci: downgrade Chromium to a version that does not cause flakes

There seems to be some issue that causes Chrome/ChromeDriver to
unexpectedly reload during the aio e2e tests, causing flakes. It is not
clear what exactly is causing the reloading, but to the best of my
knowledge it is something inside Chrome or ChromeDriver.

Pinning Chrome to r494239 (between 62.0.3185.0 and 62.0.3186.0) fixes
the flakes.

Fixes #20159

mary-poppins commented:
You can preview b6c2ea4 at https://pr20403-b6c2ea4.ngbuilds.io/.

mary-poppins commented:
You can preview 1d22df6 at https://pr20403-1d22df6.ngbuilds.io/.

mary-poppins commented:
You can preview e5bf131 at https://pr20403-e5bf131.ngbuilds.io/.

mary-poppins commented:
You can preview ffef00a at https://pr20403-ffef00a.ngbuilds.io/.

mary-poppins commented:
You can preview b77345d at https://pr20403-b77345d.ngbuilds.io/.

mary-poppins commented:
You can preview 433d00f at https://pr20403-433d00f.ngbuilds.io/.

mary-poppins commented:
You can preview a8b9261 at https://pr20403-a8b9261.ngbuilds.io/.

mary-poppins commented:
You can preview f365241 at https://pr20403-f365241.ngbuilds.io/.

mary-poppins commented:
You can preview 603e6fb at https://pr20403-603e6fb.ngbuilds.io/.

mary-poppins commented:
You can preview d0f4ec4 at https://pr20403-d0f4ec4.ngbuilds.io/.

mary-poppins commented:
You can preview a9de4e9 at https://pr20403-a9de4e9.ngbuilds.io/.

mary-poppins commented:
You can preview 776c501 at https://pr20403-776c501.ngbuilds.io/.

gkalpak commented:
The Travis failures are unrelated to this PR (and should be fixed by #20940).
The CircleCI failures are also unrelated to this PR and I have no idea what is causing them 😕

mary-poppins commented:
You can preview 462ff97 at https://pr20403-462ff97.ngbuilds.io/.

mary-poppins commented:
You can preview cd81e0e at https://pr20403-cd81e0e.ngbuilds.io/.

mary-poppins commented:
You can preview 36f5bff at https://pr20403-36f5bff.ngbuilds.io/.

mary-poppins commented:
You can preview 073a3c9 at https://pr20403-073a3c9.ngbuilds.io/.

IgorMinar commented:
Thank you, George!!!

On Mon, Dec 11, 2017 at 2:41 PM Mary Poppins <notifications@github.com>
wrote:

> You can preview 073a3c9
> <https://github.com/angular/angular/commit/073a3c9ff3f89bb27d3ba4dbbf121bf349e623e1>
> at https://pr20403-073a3c9.ngbuilds.io/.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/20403#issuecomment-350882642>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANM6ElfLgSf5_487d07TQoBMYf0cp44ks5s_a-lgaJpZM4QcbaZ>
> .
>

mary-poppins commented:
You can preview a031137 at https://pr20403-a031137.ngbuilds.io/.

mary-poppins commented:
You can preview b2a3694 at https://pr20403-b2a3694.ngbuilds.io/.

mary-poppins commented:
You can preview 4fd7e28 at https://pr20403-4fd7e28.ngbuilds.io/.

gkalpak committed:
test(aio): disable `DocViewer` animations during e2e tests

mary-poppins commented:
You can preview 19c8d21 at https://pr20403-19c8d21.ngbuilds.io/.

mary-poppins commented:
You can preview 3960b55 at https://pr20403-3960b55.ngbuilds.io/.

gkalpak commented:
I added an extra commit to disable `DocViewer` animations during e2e tests. Should be ready to go.
@petebacondarwin, @IgorMinar: PTAL

petebacondarwin reviewed (commented):
still LGTM

gkalpak commented:
Apparently, PullApprove needs you to re-approve this 🤔

petebacondarwin commented:
I think it is @igorminar that needs to approve this

gkalpak commented:
Yeah, he needs too (for `build-and-ci` and `root`). But even the `angular.io` group appeared unapproved before.

IgorMinar commented:
reapproved
